File: frontend/src/types/graph.ts
export type GraphNode = {
  id: string;
  type: string;
  label: string;
  properties: Record<string, unknown>;
};

export type GraphEdge = {
  id: string;
  source: string;
  target: string;
  type: string;
  properties: Record<string, unknown>;
};

export type GraphResponse = {
  nodes: GraphNode[];
  edges: GraphEdge[];
};

export type CentralityResult = {
  entity_id: string;
  entity_type: string;
  name: string;
  score: number;
  degree?: number | null;
};

export type Community = {
  id: number;
  size: number;
  members: Array<Record<string, unknown>>;
  key_entities: Array<Record<string, unknown>>;
};

export type CommunitiesResponse = {
  communities: Community[];
};

export type NetworkSummary = {
  total_entities: number;
  total_relationships: number;
  number_of_communities: number;
  largest_community: number;
  highest_degree_entity: Record<string, unknown> | null;
  highest_betweenness_entity: Record<string, unknown> | null;
  graph_density: number;
};

export type BridgeEntity = {
  entity: Record<string, unknown>;
  communities_connected: number[];
  bridge_score: number;
  supporting_relationships: string[];
};

export type PathResponse = { source: string; target: string; path_length: number | null; nodes: GraphNode[]; relationships: GraphEdge[]; path: string[] | null; explanation: string };
export type RelationshipEvidence = { relationship: GraphEdge | null; evidence_sources: Array<Record<string, unknown>>; source_documents: Array<Record<string, unknown>>; confidence: number | null; timestamps: string[] };

File: backend/app/schemas/graph.py
from typing import Any

from pydantic import BaseModel, Field


class GraphNodeResponse(BaseModel):
    id: str
    type: str
    label: str
    properties: dict[str, Any] = Field(default_factory=dict)


class GraphEdgeResponse(BaseModel):
    id: str
    source: str
    target: str
    type: str
    properties: dict[str, Any] = Field(default_factory=dict)


class GraphResponse(BaseModel):
    nodes: list[GraphNodeResponse]
    edges: list[GraphEdgeResponse]


class EntityResponse(GraphNodeResponse):
    case_ids: list[str] = Field(default_factory=list)
    relationship_count: int


class PathResponse(BaseModel):
    source: str
    target: str
    path_length: int | None
    nodes: list[GraphNodeResponse] = Field(default_factory=list)
    relationships: list[GraphEdgeResponse] = Field(default_factory=list)
    path: list[str] | None = None
    explanation: str


class RelationshipEvidenceResponse(BaseModel):
    relationship: GraphEdgeResponse | None
    evidence_sources: list[dict[str, Any]] = Field(default_factory=list)
    source_documents: list[dict[str, Any]] = Field(default_factory=list)
    confidence: float | None = None
    timestamps: list[str] = Field(default_factory=list)
